In [2]:
#importing decision tree, logistic regression and KNN

from sklearn.linear_model import LogisticRegression

from sklearn.tree import DecisionTreeClassifier

from sklearn.neighbors import KNeighborsClassifier

In [3]:
# importing required packages and libraries 
import pandas as pd
import numpy as np

# Part A

In [4]:
df_heart = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Heart_s.csv")

In [5]:
#priting the df for my knowledge
df_heart

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
1     67      f  asymptomatic     160   286        2    108      1.5   
2     67      f  asymptomatic     120   229        2    129      2.6   
3     37      f    nonanginal     130   250        0    187      3.5   
4     41      m    nontypical     130   204        2    172      1.4   
..   ...    ...           ...     ...   ...      ...    ...      ...   
296   45      f       typical     110   264        0    132      1.2   
297   68      f  asymptomatic     144   193        0    141      3.4   
298   57      f  asymptomatic     130   131        0    115      1.2   
299   57      m    nontypical     130   236        2    174      0.0   
300   38      f    nonanginal     138   175        0    173      0.0   

           Thal  AHD  
0         fixed   No  
1        normal  Yes  
2    reversable  Yes  
3        normal   No  
4        normal   No  
..          ...  ...  
296  reversable  Yes  
297  reversable  Yes  
298  reversable  Yes  
299      normal  Yes  
300      normal   No  

[301 rows x 10 columns]

# Part C

In [6]:
#creating python list of feature names (non catagorical)
feature_cols = ['Age' , 'RestBP', 'Chol', 'RestECG', 'MaxHR', 'Oldpeak']
#use above list to select features from original dataframe
X = df_heart[feature_cols]
#printing first 5 rows
X.head()

Age  RestBP  Chol  RestECG  MaxHR  Oldpeak
0   63     145   233        2    150      2.3
1   67     160   286        2    108      1.5
2   67     120   229        2    129      2.6
3   37     130   250        0    187      3.5
4   41     130   204        2    172      1.4

In [7]:
#selecting last column from dataframe
y = df_heart['AHD']
#checking label vector by printing every 20 values
y[::20]

0       No
20      No
40     Yes
60     Yes
80      No
100     No
120    Yes
140    Yes
160    Yes
180    Yes
200     No
220     No
240     No
260    Yes
280    Yes
300     No
Name: AHD, dtype: object


# Part D

In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=6)

# Part E

In [9]:
#KNN k=3
k=3
heart = KNeighborsClassifier(n_neighbors=k)

In [10]:
#using the fit method to train the data and labels for KNN
heart.fit(X,y)

KNeighborsClassifier(n_neighbors=3)

In [11]:
#training on training set for KNN
heart.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [12]:
#testing on testing set for KNN
y_predict = heart.predict(X_test)
print(y_predict)

['Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No'
 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'Yes'
 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No'
 'No' 'No' 'No' 'Yes' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No'
 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No'
 'Yes' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No']


In [13]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_predict)

print(accuracy)

0.6447368421052632


In [14]:
#DecisionTree and Logistic Regression
LR = LogisticRegression()
DT = DecisionTreeClassifier(random_state=5)


In [15]:
#training LR and DT models
LR.fit(X,y)
DT.fit(X,y)

DecisionTreeClassifier(random_state=5)

In [16]:
#training the models
LR.fit(X_train,y_train)
DT.fit(X_train, y_train)

DecisionTreeClassifier(random_state=5)

In [17]:
#Testing data on testing set
y_predict1 = LR.predict(X_test)
y_predict2 = DT.predict(X_test)

In [18]:
#logistic regression with random_state=6
accuracy_LR = accuracy_score(y_test, y_predict1)
print(accuracy_LR)

accuracy_DT = accuracy_score(y_test, y_predict2)
print(accuracy_DT)


0.6710526315789473
0.618421052631579


# Which one is the best?
KNN, k = 3 has a 64.5% accuracy
LR, has a 67% accuracy
DT, random_state=5 has a 61.8% accuracy.
From the numbers that I have recieved it seems that logistic regression is the most accurate model with an accuracy of 67%

# Part F

In [19]:
df_heart

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
1     67      f  asymptomatic     160   286        2    108      1.5   
2     67      f  asymptomatic     120   229        2    129      2.6   
3     37      f    nonanginal     130   250        0    187      3.5   
4     41      m    nontypical     130   204        2    172      1.4   
..   ...    ...           ...     ...   ...      ...    ...      ...   
296   45      f       typical     110   264        0    132      1.2   
297   68      f  asymptomatic     144   193        0    141      3.4   
298   57      f  asymptomatic     130   131        0    115      1.2   
299   57      m    nontypical     130   236        2    174      0.0   
300   38      f    nonanginal     138   175        0    173      0.0   

           Thal  AHD  
0         fixed   No  
1        normal  Yes  
2    reversable  Yes  
3        normal   No  
4        normal   No  
..          ...  ...  
296  reversable  Yes  
297  reversable  Yes  
298  reversable  Yes  
299      normal  Yes  
300      normal   No  

[301 rows x 10 columns]

In [20]:
#creating dummies for gender
gender = pd.get_dummies(df_heart.Gender)
gender

f  m
0    1  0
1    1  0
2    1  0
3    1  0
4    0  1
..  .. ..
296  1  0
297  1  0
298  1  0
299  0  1
300  1  0

[301 rows x 2 columns]

In [21]:
#creating dummies for chestpain
chestpain = pd.get_dummies(df_heart.ChestPain)
chestpain

asymptomatic  nonanginal  nontypical  typical
0               0           0           0        1
1               1           0           0        0
2               1           0           0        0
3               0           1           0        0
4               0           0           1        0
..            ...         ...         ...      ...
296             0           0           0        1
297             1           0           0        0
298             1           0           0        0
299             0           0           1        0
300             0           1           0        0

[301 rows x 4 columns]

In [22]:
#creating dummies for Thal
thal = pd.get_dummies(df_heart.Thal)
thal

fixed  normal  reversable
0        1       0           0
1        0       1           0
2        0       0           1
3        0       1           0
4        0       1           0
..     ...     ...         ...
296      0       0           1
297      0       0           1
298      0       0           1
299      0       1           0
300      0       1           0

[301 rows x 3 columns]

In [23]:
#merging all dummy columns to original dataframe
merged = pd.concat([df_heart,gender,chestpain,thal],axis = 'columns')
merged

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
1     67      f  asymptomatic     160   286        2    108      1.5   
2     67      f  asymptomatic     120   229        2    129      2.6   
3     37      f    nonanginal     130   250        0    187      3.5   
4     41      m    nontypical     130   204        2    172      1.4   
..   ...    ...           ...     ...   ...      ...    ...      ...   
296   45      f       typical     110   264        0    132      1.2   
297   68      f  asymptomatic     144   193        0    141      3.4   
298   57      f  asymptomatic     130   131        0    115      1.2   
299   57      m    nontypical     130   236        2    174      0.0   
300   38      f    nonanginal     138   175        0    173      0.0   

           Thal  AHD  f  m  asymptomatic  nonanginal  nontypical  typical  \
0         fixed   No  1  0             0           0           0        1   
1        normal  Yes  1  0             1           0           0        0   
2    reversable  Yes  1  0             1           0           0        0   
3        normal   No  1  0             0           1           0        0   
4        normal   No  0  1             0           0           1        0   
..          ...  ... .. ..           ...         ...         ...      ...   
296  reversable  Yes  1  0             0           0           0        1   
297  reversable  Yes  1  0             1           0           0        0   
298  reversable  Yes  1  0             1           0           0        0   
299      normal  Yes  0  1             0           0           1        0   
300      normal   No  1  0             0           1           0        0   

     fixed  normal  reversable  
0        1       0           0  
1        0       1           0  
2        0       0           1  
3        0       1           0  
4        0       1           0  
..     ...     ...         ...  
296      0       0           1  
297      0       0           1  
298      0       0           1  
299      0       1           0  
300      0       1           0  

[301 rows x 19 columns]

In [24]:
#dropping original catagorical data
final = merged.drop(['Gender', 'ChestPain', 'Thal'],axis = 'columns')
final

Age  RestBP  Chol  RestECG  MaxHR  Oldpeak  AHD  f  m  asymptomatic  \
0     63     145   233        2    150      2.3   No  1  0             0   
1     67     160   286        2    108      1.5  Yes  1  0             1   
2     67     120   229        2    129      2.6  Yes  1  0             1   
3     37     130   250        0    187      3.5   No  1  0             0   
4     41     130   204        2    172      1.4   No  0  1             0   
..   ...     ...   ...      ...    ...      ...  ... .. ..           ...   
296   45     110   264        0    132      1.2  Yes  1  0             0   
297   68     144   193        0    141      3.4  Yes  1  0             1   
298   57     130   131        0    115      1.2  Yes  1  0             1   
299   57     130   236        2    174      0.0  Yes  0  1             0   
300   38     138   175        0    173      0.0   No  1  0             0   

     nonanginal  nontypical  typical  fixed  normal  reversable  
0             0           0        1      1       0           0  
1             0           0        0      0       1           0  
2             0           0        0      0       0           1  
3             1           0        0      0       1           0  
4             0           1        0      0       1           0  
..          ...         ...      ...    ...     ...         ...  
296           0           0        1      0       0           1  
297           0           0        0      0       0           1  
298           0           0        0      0       0           1  
299           0           1        0      0       1           0  
300           1           0        0      0       1           0  

[301 rows x 16 columns]

# Part G

In [25]:
#creating python list of feature names (non catagorical)
features = ['Age' , 'RestBP', 'Chol', 'RestECG', 'MaxHR', 'Oldpeak','f','m','asymptomatic','nonanginal','nontypical','typical','fixed','normal','reversable']
#use above list to select features from original dataframe
X1 = final[features]
#printing first 5 rows
X1.head()

Age  RestBP  Chol  RestECG  MaxHR  Oldpeak  f  m  asymptomatic  nonanginal  \
0   63     145   233        2    150      2.3  1  0             0           0   
1   67     160   286        2    108      1.5  1  0             1           0   
2   67     120   229        2    129      2.6  1  0             1           0   
3   37     130   250        0    187      3.5  1  0             0           1   
4   41     130   204        2    172      1.4  0  1             0           0   

   nontypical  typical  fixed  normal  reversable  
0           0        1      1       0           0  
1           0        0      0       1           0  
2           0        0      0       0           1  
3           0        0      0       1           0  
4           1        0      0       1           0

In [26]:
#selecting AHD in data frame as y
y1 = final['AHD']
y1

0       No
1      Yes
2      Yes
3       No
4       No
      ... 
296    Yes
297    Yes
298    Yes
299    Yes
300     No
Name: AHD, Length: 301, dtype: object

In [26]:
X_trainf, X_testf, y_trainf, y_testf = train_test_split(X1, y1, test_size=0.25, random_state=6)

In [27]:
#KNN on new dataset
k = 3
fin = KNeighborsClassifier(n_neighbors=k) 

In [28]:
#fitting data for knn
fin.fit(X1,y1)

KNeighborsClassifier(n_neighbors=3)

In [29]:
#Training ONLY on training set
fin.fit(X_trainf,y_trainf)

KNeighborsClassifier(n_neighbors=3)

In [30]:
y_predictf = fin.predict(X_testf)

In [31]:
acc = accuracy_score(y_testf,y_predictf)
acc

0.6447368421052632

In [32]:
#creating objects for DT and LR
my_logreg = LogisticRegression()
my_decisiontree = DecisionTreeClassifier(random_state=5)

In [33]:
my_decisiontree.fit(X1,y1)

DecisionTreeClassifier(random_state=5)

In [43]:
#changed max iterations to avoid errors and convergence issues
my_logreg.fit(X1,y1)
my_logreg=LogisticRegression(max_iter=10000)

In [44]:
#training on training set
my_logreg.fit(X_trainf,y_trainf)
my_decisiontree.fit(X_trainf,y_trainf)

DecisionTreeClassifier(random_state=5)

In [45]:
# Testing on the testing set:

y_predict_lr = my_logreg.predict(X_testf)

y_predict_dt = my_decisiontree.predict(X_testf)


In [46]:
score_lr = accuracy_score(y_test, y_predict_lr)
score_dt = accuracy_score(y_test, y_predict_dt)

print(score_lr)
print(score_dt)

0.7763157894736842
0.7368421052631579


# How does the prediction accuracy change for each method?
The new data inputs seemed to increase the accuracy a little. KNN accuracy stayed the same, LR accuracy increased by 10% and DT accuracy increased by 12%.   

# Part H

In [47]:
# importing cross validation method:
from sklearn.model_selection import cross_val_score

In [54]:
#using same features as part F
features = ['Age' , 'RestBP', 'Chol', 'RestECG', 'MaxHR', 'Oldpeak','f','m','asymptomatic','nonanginal','nontypical','typical','fixed','normal','reversable']

X = final[features]

y = final['AHD']

X

Age  RestBP  Chol  RestECG  MaxHR  Oldpeak  f  m  asymptomatic  \
0     63     145   233        2    150      2.3  1  0             0   
1     67     160   286        2    108      1.5  1  0             1   
2     67     120   229        2    129      2.6  1  0             1   
3     37     130   250        0    187      3.5  1  0             0   
4     41     130   204        2    172      1.4  0  1             0   
..   ...     ...   ...      ...    ...      ... .. ..           ...   
296   45     110   264        0    132      1.2  1  0             0   
297   68     144   193        0    141      3.4  1  0             1   
298   57     130   131        0    115      1.2  1  0             1   
299   57     130   236        2    174      0.0  0  1             0   
300   38     138   175        0    173      0.0  1  0             0   

     nonanginal  nontypical  typical  fixed  normal  reversable  
0             0           0        1      1       0           0  
1             0           0        0      0       1           0  
2             0           0        0      0       0           1  
3             1           0        0      0       1           0  
4             0           1        0      0       1           0  
..          ...         ...      ...    ...     ...         ...  
296           0           0        1      0       0           1  
297           0           0        0      0       0           1  
298           0           0        0      0       0           1  
299           0           1        0      0       1           0  
300           1           0        0      0       1           0  

[301 rows x 15 columns]

In [55]:
#setting k = 3 for KNN
k = 3
my_KNN = KNeighborsClassifier(n_neighbors=k)

In [56]:
#creating a list for cross val scores
accuracy_KNN = cross_val_score(my_KNN, X, y, cv=10, scoring='accuracy')

print(accuracy_KNN)

[0.70967742 0.63333333 0.56666667 0.66666667 0.6        0.5
 0.66666667 0.7        0.56666667 0.73333333]


In [57]:
#Printing accuracy for CV on KNN where K = 3
accuracy_K = accuracy_KNN.mean()

print(accuracy_K)

0.6343010752688172


In [58]:
#Applying CV with DT and random state = 5
my_DT = DecisionTreeClassifier(random_state=5)
#creating object of DT to hold CV scores
accuracy_DT = cross_val_score(my_DT, X, y, cv=10, scoring='accuracy')
#printing CV scores
print(accuracy_DT)

[0.77419355 0.76666667 0.8        0.76666667 0.8        0.7
 0.7        0.66666667 0.7        0.83333333]


In [59]:
#Creating varibale to hold mean of all DT accuracies 
accuracyD = accuracy_DT.mean()

print(accuracyD)

0.750752688172043


In [60]:
#starting to apply CV with logistic regression
my_LR = LogisticRegression(max_iter = 10000)
#creating list to store cv values from LR
accuracy_LR = cross_val_score(my_LR, X, y, cv = 10, scoring = 'accuracy')
#printing LR CV score list
print(accuracy_LR)

[0.77419355 0.8        0.83333333 0.86666667 0.93333333 0.7
 0.76666667 0.8        0.76666667 0.83333333]


In [61]:
#creating object to store mean of LR score values
accuracyL = accuracy_LR.mean()
#printing out accuracy of LR and CV 
print(accuracyL)

0.8074193548387096
